src/builder/create_attachment.rs

```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+}
+
+/// You can add new attachments and edit existing ones using this builder.
+///
+/// When this builder is _not_ supplied in a message edit, Discord keeps the attachments intact.
+/// However, as soon as a builder is supplied, Discord removes all attachments from the message.
+/// If you want to keep old attachments, you have to specify this for each attachment using
+/// [`Self.keep()`].
+///
+/// # Examples
+///
+/// ## Removing all attachments
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(EditAttachments::new())).await?;
+/// # Ok(()) }
+/// ```
+///
+/// ## Adding a new attachment without deleting existing attachments
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message, my_attachment: CreateAttachment) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new_keep_all(&msg).add(my_attachment)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// ## Delete all but the first attachment
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message, my_attachment: CreateAttachment) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new().keep(msg.attachments[0].id)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// ## Delete only the first attachment
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message, my_attachment: CreateAttachment) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new_keep_all(&msg).dont_keep(msg.attachments[0].id)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// # Notes
+///
+/// Internally, this type is used not just for message editing endpoints, but also for message
+/// creation endpoints.
+#[derive(Default, Debug, Clone, serde::Serialize)]
+#[serde(transparent)]
+#[must_use]
+pub struct EditAttachments {
+    new_and_existing_attachments: Vec<NewOrExisting>,
+}
+
+impl EditAttachments {
+    /// An empty attachments builder.
+    ///
+    /// Existing attachments are not kept by default, either. See [`Self::new_keep_all()`] or
+    /// [`Self::keep()`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a new attachments builder that keeps all existing attachments.
+    ///
+    /// Shorthand for [`Self::new()`] and calling [`Self::keep()`] for every
+    /// [`AttachmentId`] in [`Message::attachments`].
+    ///
+    /// If you only want to keep a subset of attachments from the message, either implement this
+    /// method manually, or use [`Self::dont_keep()`].
+    ///
+    /// **Note: this EditAttachments must be run on the same message as is supplied here, or else
+    /// Discord will throw an error!**
+    pub fn new_keep_all(msg: &Message) -> Self {
```

Comment:
The `new_` prefix is unnecessary. This could also be named `retain_all`, similar to `keep` below. What do you think?

Reply:
I went through a few iterations on the method name and deliberately decided to keep `new`, to make clear that this is a constructor method.

Reply:
That's true, but the signature is pretty short and the absence of `self` I think is easy enough to notice. Moreover, I think having the name be consistent with `keep` is important.

Reply:
Alright